overlord/snapstate/snapmgr.go

```diff
@@ -215,6 +216,55 @@ func (compsu *ComponentSetup) Revision() snap.Revision {
 	return compsu.CompSideInfo.Revision
 }
 
+// ComponentSetupFromSnapSetup returns a list of ComponentSetup structs for the
+// given task. Since the task could originate from one of a few different
+// scenarios, we inspect the task for various keys to determine how to find the
+// component setups.
+//
+// The task could originate from:
+// * Installing a singular component for an already installed snap
+// * Installing/refreshing a snap with components
+// * Installing/refreshing a snap without any components
+func ComponentSetupsForTask(t *state.Task) ([]*ComponentSetup, error) {
+	switch {
+	case t.Has("component-setup") || t.Has("component-setup-task"):
+		// task comes from a component installation
+		compsup, _, err := TaskComponentSetup(t)
+		if err != nil {
+			return nil, err
+		}
+		return []*ComponentSetup{compsup}, nil
+	case t.Has("component-setups") || t.Has("component-setups-task"):
+		// TODO: test this branch once we know more about refreshing/installing
+		// snaps with components
```

Comment:
not an immediate concern for this PR by itself but we need to be careful with the code that was added here:

https://github.com/snapcore/snapd/pull/13842/files#diff-5c859073a3ee961433aaf8b56df66f8000494e5174108e7b1558699af41baa1cR414

maybe we should leave comment nearby on this. But if we have a task set that operates both on a snap and components at the same time and includes setup-profiles, we will need to attach component-setups to the snap-setup carrying task for the snap, but still separate add component-setup  to the prepare-component or equivalent task for each component.  Does this make sense?